File: AtCoder/ABC175/D/d.py
def main():
    import numpy as np
    N, K = map(int, input().split())
    P = [int(x) - 1 for x in input().split()]
    P = np.array(P, np.int64)
    C = np.array(input().split(), np.int64)
    ans = -(10 ** 18)
    for i in range(N):
        j = i
        S = 0
        L = 0
        for _ in range(K):
            j = P[j]
            S += C[j]
            L += 1
            ans = max(ans, S)
            if j == i:
                break
        if j == i:
            M = max(0, K // L - 1)
            T = S * M
            for _ in range(M * L, K):
                j = P[j]
                T += C[j]
                ans = max(ans, T)
    print(ans)


if __name__ == "__main__":
  main()

File: AtCoder/ABC175/D/d.rs
extern crate proconio;
use proconio::input;

fn main() {
    input! {
        n: usize,
        k: usize,
        mut p: [usize; n],
        c: [i64; n],
    }
    let p = p.iter().map(|x| x - 1).collect::<Vec<_>>();
    let mut q = vec![0; n];
    for (i, &x) in p.iter().enumerate() {
        q[x] = i;
    }
    let mut ans = std::i64::MIN;
    for i in 0..n {
        let mut cur = i;
        let mut s = 0;
        let mut cycle = false;
        let mut len: usize = 0;
        for _ in 0..k {
            cur = p[cur];
            s += c[cur];
            ans = std::cmp::max(ans, s);
            len += 1;
            if cur == i {
                cycle = true;
                break;
            }
        }
        if cycle {
            let mut a = k / len;
            if a >= 2 {
                a -= 2;
            }
            s = s * a as i64;
            cur = i;
            let mut t = 0;
            for _ in 0..(k - a * len) {
                cur = p[cur];
                t += c[cur];
                ans = std::cmp::max(ans, s + t);
            }
        }
    }
    println!("{}", ans);
}
